examples/core/dof/app.js

```diff
@@ -0,0 +1,433 @@
+          boxMatrices.set(boxXforms[boxI].matrix, boxI * 16);
+        ++boxI;
+        }
+    }
+
+
+    ////////////////////////////////////////////////
+    // Set up frambuffers. 
+    ////////////////////////////////////////////////
+
+    // Need to ensure both color and depth targets can be sampled.
+    const mainFramebuffer = new Framebuffer(gl, {
+      width: gl.drawingBufferWidth,
+      height: gl.drawingBufferHeight,
+      attachments: {
+        [GL.COLOR_ATTACHMENT0]: new Texture2D(gl, {
+          format: GL.RGBA,
+          type: GL.UNSIGNED_BYTE,
+          width: gl.drawingBufferWidth,
+          height: gl.drawingBufferHeight,
+          mipmaps: false,
+          parameters: {
+            [GL.TEXTURE_MIN_FILTER]: GL.LINEAR,
+            [GL.TEXTURE_MAG_FILTER]: GL.LINEAR,
+            [GL.TEXTURE_WRAP_S]: GL.CLAMP_TO_EDGE,
+            [GL.TEXTURE_WRAP_T]: GL.CLAMP_TO_EDGE
+          }
+        }),
+        [GL.DEPTH_ATTACHMENT]: new Texture2D(gl, {
+          format: GL.DEPTH_COMPONENT16,
+          type: GL.UNSIGNED_SHORT,
+          dataFormat: GL.DEPTH_COMPONENT,
+          width: gl.drawingBufferWidth,
+          height: gl.drawingBufferHeight,
+          mipmaps: false,
+          parameters: {
+            [GL.TEXTURE_MIN_FILTER]: GL.NEAREST,
+            [GL.TEXTURE_MAG_FILTER]: GL.NEAREST,
+            [GL.TEXTURE_WRAP_S]: GL.CLAMP_TO_EDGE,
+            [GL.TEXTURE_WRAP_T]: GL.CLAMP_TO_EDGE
+          }
+        })
+      }
+    })
+
+    // Postprocessing FBO doesn't need a depth attachment.
+    const dofFramebuffer = new Framebuffer(gl, { width: gl.drawingBufferWidth, height: gl.drawingBufferHeight, depth: false});
+
+    return loadTextures(gl, {
+      urls: ['webgl-logo.png'],
+      mipmaps: true,
+      parameters: {
+        [gl.TEXTURE_MAG_FILTER]: gl.LINEAR,
+        [gl.TEXTURE_MIN_FILTER]: gl.LINEAR_MIPMAP_NEAREST
+      }
+    }).then((textures) => {
+
+      ////////////////////////////////////////////
+      // Set up instanced model to draw boxes.
+      ////////////////////////////////////////////
+
+      const cubeGeo = new CubeGeometry().getAttributes();
+      const positionBuffer = new Buffer(gl, cubeGeo.positions.value);
+      const normalBuffer = new Buffer(gl, cubeGeo.normals.value);
+      const uvBuffer = new Buffer(gl, cubeGeo.texCoords.value);
+      const indexBuffer = new Buffer(gl, { data: cubeGeo.indices.value, target: gl.ELEMENT_ARRAY_BUFFER });
+
+      // This containes the instanced matrix attributes
+      let matrixBuffer = new Buffer(gl, boxMatrices);
+
+      const instancedBoxes = new Model(gl, { 
+        program: sceneProgram,
+        attributes: {
+          positions: positionBuffer,
+          normals: normalBuffer,
+          uvs: uvBuffer,
+
+          // Attributes are limited to 4 components,
+          // So we have to split the matrices across
+          // 4 attributes. They're reconstructed in 
+          // the vertex shader.
+          modelMatCol1: {
```

Comment:
Looks like complicated setup. Can this be improved? Looks like all you need is position offset so each instance can be placed at a different offset, rest of the transformations can be constant (like rotation etc), so can you make a new `instancePositions` attribute and set rest of the matrix as uniform?

Reply:
There's a translation as well as animating rotations about the X and Z axes, so that would be a different animation. More generally, the idea here is to show how to send an arbitrary transform matrix per instance. I also added a note to the PR description about how it might be worth it for Luma to support doing this directly through the API.